### PR TITLE
Yihui/api demo

### PR DESCRIPTION
**Pull request by YihuiGuo**
Description:
Need to build and publish UI-based user experience, consisting of a Restful backend and a frontend UI.
Backend design doc:
https://microsoft.sharepoint.com/:w:/t/AICustomerEngineeringTeam/ESCFZIcBrpxBmkITXDwT-7EB9dB_V9F-eStOqcFPh1ZfOg

### CODE REVIEW

**Comment by YihuiGuo** [comment]
Add link to design document : https://microsoft.sharepoint.com/:w:/t/AICustomerEngineeringTeam/ESCFZIcBrpxBmkITXDwT-7EB9dB_V9F-eStOqcFPh1ZfOg

**Comment by hangfei** [comment]
Please follow: https://github.com/linkedin/feathr/blob/main/docs/dev_guide/pull_request_guideline.md

**Comment by xiaoyongzhu** [comment]
@YihuiGuo can you merge main to your repo so the test can run?

**Comment by YihuiGuo** [comment]
Linked Issue : https://github.com/linkedin/feathr/issues/169

**Comment by YihuiGuo** [comment]
> 

Already rebased onto main branch

**Comment by YihuiGuo** [comment]
> @YihuiGuo can you merge main to your repo so the test can run?

Already rebased onto main branch

File: feathr_project/feathr/api/app/backends/empty_backend_impl.py
Comment (xiaoyongzhu) [inline feathr_project/feathr/api/app/backends/empty_backend_impl.py:6]: What's this for?

File: feathr_project/feathr/api/app/core/configs.py
Comment (xiaoyongzhu) [inline feathr_project/feathr/api/app/core/configs.py:6]: document it to make sure folks know how to get it

File: feathr_project/feathr/api/app/core/configs.py
Comment (xiaoyongzhu) [inline feathr_project/feathr/api/app/core/configs.py:8]: don't use "aa"

File: feathr_project/feathr/api/app/core/error_code.py
Comment (xiaoyongzhu) [inline feathr_project/feathr/api/app/core/error_code.py:4]: I thought REST API have a lot error codes?

File: feathr_project/feathr/api/app/routes/project.py
Comment (xiaoyongzhu) [inline feathr_project/feathr/api/app/routes/project.py:11]: Seems I cannot comment on empty files. Are offline_faeture.py and online_faeture.py required? Looks like they are emtpy files

File: feathr_project/feathr/api/requirements.txt
Comment (xiaoyongzhu) [inline feathr_project/feathr/api/requirements.txt:8]: use main?

File: dockerfile
Comment (xiaoyongzhu) [inline dockerfile:17]: is root folder the right place for this file?

File: feathr_project/feathr/_feature_registry.py
Comment (xiaoyongzhu) [inline feathr_project/feathr/_feature_registry.py:739]: should we call it GUID or something? to make it more clear.

File: feathr_project/feathr/_feature_registry.py
Comment (xiaoyongzhu) [inline feathr_project/feathr/_feature_registry.py:881]: delete print?

File: feathr_project/feathr/api/app/backends/empty_backend_impl.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/backends/empty_backend_impl.py:6]: Seems not need to use this for now, will be deteled

File: feathr_project/feathr/api/app/core/configs.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/core/configs.py:8]: Changed to "mockAppServiceKey"

File: feathr_project/feathr/api/requirements.txt
Comment (YihuiGuo) [reply feathr_project/feathr/api/requirements.txt:8]: updated

File: feathr_project/feathr/_feature_registry.py
Comment (YihuiGuo) [reply feathr_project/feathr/_feature_registry.py:881]: Ok, done

File: feathr_project/feathr/_feature_registry.py
Comment (YihuiGuo) [reply feathr_project/feathr/_feature_registry.py:739]: My suggestion is we keep it as it is for now, since changing this might break Blair's UI...  I can change this before his PR merges

File: feathr_project/feathr/api/app/core/configs.py
Comment (xiaoyongzhu) [reply feathr_project/feathr/api/app/core/configs.py:4]: let's use V1

File: feathr_project/feathr/api/app/core/error_code.py
Comment (xiaoyongzhu) [reply feathr_project/feathr/api/app/core/error_code.py:4]: Can you explain a bit on how it works? I thought HTTP also has something representing invalid code etc.

File: feathr_project/feathr/api/app/core/configs.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/core/configs.py:4]: Ok

File: dockerfile
Comment (YihuiGuo) [reply dockerfile:17]: @blrchen  will be working on dockerize the api, and will fix later

File: feathr_project/feathr/api/app/core/configs.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/core/configs.py:6]: OK, will update the design doc.

File: feathr_project/feathr/api/app/core/configs.py
Comment (xiaoyongzhu) [reply feathr_project/feathr/api/app/core/configs.py]: sorry I meant v1 in lower case

File: feathr_project/feathr/api/app/backends/backend_base.py
Comment (hangfei) [inline feathr_project/feathr/api/app/backends/backend_base.py:4]: class doc

File: dockerfile
Comment (hangfei) [inline dockerfile:12]: change the folder name to feature_registry_api

File: feathr_project/feathr/api/app/backends/backend_base.py
Comment (hangfei) [inline feathr_project/feathr/api/app/backends/backend_base.py:4]: add documentation for each method. if not needed for now, just remove them for now and u can always add it back in future PRs.

File: feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py
Comment (hangfei) [inline feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py:9]: is this coming from our repo? it should starts with feathr.x.y

File: feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py
Comment (hangfei) [inline feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py:12]: this should always come from yaml config and we should not hard code our test env stuff here.

File: feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py
Comment (hangfei) [inline feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py:15]: class doc.

File: feathr_project/feathr/api/app/backends/purview_backend/purview_backend_config.py
Comment (hangfei) [inline feathr_project/feathr/api/app/backends/purview_backend/purview_backend_config.py:6]: what's this config_path?

File: feathr_project/feathr/api/app/core/error_handling.py
Comment (hangfei) [inline feathr_project/feathr/api/app/core/error_handling.py:17]: please state what happened(getting 404 from backends etc), direction reason(you already have) and potential solutions or what usres should do.

File: feathr_project/feathr/api/app/core/error_handling.py
Comment (hangfei) [inline feathr_project/feathr/api/app/core/error_handling.py:15]: this acn go with FeathrApiException?

File: feathr_project/setup.py
Comment (hangfei) [inline feathr_project/setup.py:10]: dont' change this. This wont' work for pypi.

File: feathr_project/feathr/api/app/routes/recommendation.py
Comment (hangfei) [inline feathr_project/feathr/api/app/routes/recommendation.py:1]: rename the file name to fit its content

File: feathr_project/setup.py
Comment (YihuiGuo) [reply feathr_project/setup.py:10]: Reverted

File: feathr_project/feathr/api/app/routes/recommendation.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/routes/recommendation.py:1]: deleted, will have new route to handle search scenarios

File: feathr_project/feathr/api/app/core/error_handling.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/core/error_handling.py:17]: Thanks, enriched description

File: feathr_project/feathr/api/app/backends/purview_backend/purview_backend_config.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/backends/purview_backend/purview_backend_config.py:6]: no longer used, deleted

File: feathr_project/feathr/api/app/backends/backend_base.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/backends/backend_base.py:4]: Changed

File: dockerfile
Comment (YihuiGuo) [reply dockerfile:12]: @blrchen will fix these docker-related modifications later

File: feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py:9]: @blrchen will fix these docker-related modifications later

File: feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py:15]: Added

File: feathr_project/feathr/api/app/backends/backend_base.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/backends/backend_base.py:4]: Added

File: feathr_project/feathr/_feature_registry.py
Comment (YihuiGuo) [reply feathr_project/feathr/_feature_registry.py:867]: This is called internally in get_feature_by_fqdn

File: feathr_project/feathr/_feature_registry.py
Comment (YihuiGuo) [reply feathr_project/feathr/_feature_registry.py:842]: expanded to fully_qualifeid_domain_name

File: feathr_project/feathr/_feature_registry.py
Comment (YihuiGuo) [reply feathr_project/feathr/_feature_registry.py:892]: will change this in later work, implementing search & fuzzy search features

File: feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py
Comment (YihuiGuo) [reply feathr_project/feathr/api/app/backends/purview_backend/purview_backend.py:12]: @blrchen can we merge these into config yaml in docker deployment?

**Review by hangfei — COMMENTED** [review]
the documentation should explain the paramters as well. please fix all of them.

File: feathr_project/feathr/api/app/routes/project.py
Comment (hangfei) [inline feathr_project/feathr/api/app/routes/project.py:16]: is the convention to just use methods and not classes?

Comment (YihuiGuo) [reply feathr_project/feathr/api/app/routes/project.py:16]: Yes, ref : https://fastapi.tiangolo.com/tutorial/bigger-applications/#import-apirouter

Restful interfaces by nature should be stateless, and I think method-based is reasonable.

**Comment by hangfei** [comment]
Please follow https://github.com/linkedin/feathr/blob/main/docs/dev_guide/update_python_docs.md to make sure your py docs are generated. Please attach a sreenshot of the docs. This can defer to next other PR if the current PR is blocking others.

**Comment by YihuiGuo** [comment]
Screen Shot 2022-05-06 at 21 40 23

**Comment by hangfei** [comment]
The documentation is broken on your end. Not sure if it's env issue or bug.

**Comment by hangfei** [comment]
LGTM. Please resolve the conflicts and i will shipit.

**Review by hangfei — APPROVED** [review]

**Review by hangfei — APPROVED** [review]

**Review by xiaoyongzhu — APPROVED** [review]

**Pull request merged**
